src/Aspire.Hosting/ApplicationModel/InteractionService.cs

```diff
@@ -0,0 +1,287 @@
+                updateEvent = _interactionCollection.Remove(interactionUpdate.Id);
+            }
+            else
+            {
+                if (_interactionCollection.Contains(interactionUpdate.Id))
+                {
+                    // Should never happen, but throw descriptive exception if it does.
+                    throw new InvalidCastException($"An interaction with ID {interactionUpdate.Id} already exists. Interaction IDs must be unique.");
+                }
+
+                _interactionCollection.Add(interactionUpdate);
+            }
+
+            OnInteractionUpdated?.Invoke(interactionUpdate);
+        }
+    }
+
+    internal void CompleteInteraction(int interactionId, Func<Interaction, InteractionResult> createResult)
+    {
+        lock (_onInteractionUpdatedLock)
+        {
+            if (_interactionCollection.TryGetValue(interactionId, out var interactionState))
+            {
+                var result = createResult(interactionState);
+
+                interactionState.TaskCompletionSource.TrySetResult(result);
+                interactionState.State = result.Canceled ? Interaction.InteractionState.Canceled : Interaction.InteractionState.Complete;
+                _interactionCollection.Remove(interactionId);
+                OnInteractionUpdated?.Invoke(interactionState);
+            }
+            else
+            {
+                throw new KeyNotFoundException($"No interaction found with ID {interactionId}.");
+            }
+        }
+    }
+
+    internal async IAsyncEnumerable<Interaction> SubscribeInteractionUpdates([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var channel = Channel.CreateUnbounded<Interaction>();
+
+        void WriteToChannel(Interaction resourceEvent) =>
+            channel.Writer.TryWrite(resourceEvent);
+
+        List<Interaction> pendingInteractions;
+
+        lock (_onInteractionUpdatedLock)
+        {
+            OnInteractionUpdated += WriteToChannel;
+
+            pendingInteractions = _interactionCollection.ToList();
+        }
+
+        foreach (var interaction in pendingInteractions)
+        {
+            yield return interaction;
+        }
+
+        try
+        {
+            await foreach (var item in channel.Reader.ReadAllAsync(cancellationToken).ConfigureAwait(false))
+            {
+                yield return item;
+            }
+        }
+        finally
+        {
+            lock (_onInteractionUpdatedLock)
+            {
+                OnInteractionUpdated -= WriteToChannel;
+            }
+
+            channel.Writer.TryComplete();
+        }
+    }
+}
+
+internal class InteractionCollection : KeyedCollection<int, Interaction>
+{
+    protected override int GetKeyForItem(Interaction item) => item.Id;
+}
+
+/// <summary>
+/// 
+/// </summary>
+public class InteractionResult
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public object? Data { get; set; }
```

Comment:
```suggestion
    public object? Data { get; }
```

Is mutation necessary here?